File: engine/sdk/ts-kit-engine/src/collection.ts
export interface Version { id: string; entityId: string; timestamp: string; data: unknown }

export interface CollectionQuery { limit?: number; offset?: number; sort?: string; search?: string }

export class Collection<T extends { id: string }> {
  constructor(private baseURL: string, private type: string) {}

  private url(path = ""): string {
    return `${this.baseURL}/${this.type}${path}`;
  }

  private async request<R>(method: string, path: string, body?: unknown): Promise<R> {
    const res = await fetch(this.url(path), {
      method,
      headers: { "Content-Type": "application/json" },
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });
    if (!res.ok) throw new Error(`${method} ${this.type}${path}: ${res.status}`);
    if (res.status === 204) return undefined as R;
    return (await res.json()) as R;
  }

  async create(data: Omit<T, "id"> | T): Promise<T> {
    return this.request<T>("POST", "/", data);
  }

  async get(id: string): Promise<T> {
    return this.request<T>("GET", `/${encodeURIComponent(id)}`);
  }

  async list(q?: CollectionQuery): Promise<T[]> {
    const p = new URLSearchParams();
    if (q?.limit !== undefined) p.set("limit", String(q.limit));
    if (q?.offset !== undefined) p.set("offset", String(q.offset));
    if (q?.sort) p.set("sort", q.sort);
    if (q?.search) p.set("search", q.search);
    const qs = p.toString();
    return this.request<T[]>("GET", qs ? `/?${qs}` : "/");
  }

  async update(id: string, data: T): Promise<T> {
    return this.request<T>("PUT", `/${encodeURIComponent(id)}`, data);
  }

  async delete(id: string): Promise<void> {
    return this.request<void>("DELETE", `/${encodeURIComponent(id)}`);
  }

  async history(id: string): Promise<Version[]> {
    return this.request<Version[]>("GET", `/${encodeURIComponent(id)}/versions`);
  }

  async revert(id: string, versionId: string): Promise<T> {
    return this.request<T>(
      "POST",
      `/${encodeURIComponent(id)}/revert/${encodeURIComponent(versionId)}`,
    );
  }
}

File: engine/sdk/py-kit-engine/kit_engine/collection.py
"""CRUD operations on a typed collection."""

from __future__ import annotations

import requests


class Collection:
    """REST client for a single kit collection type."""

    def __init__(self, base_url: str, type_name: str):
        self._url = f"{base_url}/{type_name}"

    def create(self, data: dict) -> dict:
        resp = requests.post(f"{self._url}/", json=data)
        resp.raise_for_status()
        return resp.json()

    def get(self, id: str) -> dict:
        resp = requests.get(f"{self._url}/{id}")
        resp.raise_for_status()
        return resp.json()

    def list(
        self, *, limit: int = 20, offset: int = 0, sort: str = "", search: str = ""
    ) -> list[dict]:
        params: dict = {"limit": limit, "offset": offset}
        if sort:
            params["sort"] = sort
        if search:
            params["search"] = search
        resp = requests.get(f"{self._url}/", params=params)
        resp.raise_for_status()
        return resp.json()

    def update(self, id: str, data: dict) -> dict:
        resp = requests.put(f"{self._url}/{id}", json=data)
        resp.raise_for_status()
        return resp.json()

    def delete(self, id: str) -> None:
        resp = requests.delete(f"{self._url}/{id}")
        resp.raise_for_status()

    def history(self, id: str) -> list[dict]:
        resp = requests.get(f"{self._url}/{id}/history")
        resp.raise_for_status()
        return resp.json()

    def revert(self, id: str, version_id: str) -> dict:
        resp = requests.post(f"{self._url}/{id}/revert/{version_id}")
        resp.raise_for_status()
        return resp.json()
